golang/cmd/kafka-to-postgresql/pxProcessValue.go

```diff
@@ -0,0 +1,232 @@
+							reason:      reason,
+							errorString: &errStr,
+						}
+					}
+				}
+				messages = make([]*kafka.Message, 0)
+
+				break
+			}
+		}
+	}
+	for _, message := range messages {
+		highThroughputPutBackChannel <- PutBackChanMsg{
+			msg:         message,
+			reason:      "Shutting down",
+			errorString: nil,
+		}
+	}
+}
+
+func writeProcessValueToDatabase(messages []*kafka.Message) (putBackMsg []*kafka.Message, err error, putback bool, reason string) {
+	//zap.S().Debugf("[HT][PV] Writing %d messages to database", len(messages))
+	var txn *sql.Tx = nil
+	txn, err = db.Begin()
+	if err != nil {
+		zap.S().Errorf("Error starting transaction: %s", err.Error())
+		return messages, err, true, "Error starting transaction"
+	}
+
+	//zap.S().Debugf("[HT][PV] 1")
+	{
+		stmt := txn.Stmt(statement.CreateTmpProcessValueTable64)
+
+		_, err = stmt.Exec()
+		if err != nil {
+			txn.Rollback()
+			zap.S().Errorf("Error creating temporary table: %s", err.Error())
+			return messages, err, true, "Error creating temporary table"
+		}
+	}
+	putBackMsg = make([]*kafka.Message, 0)
+	toCommit := float64(0)
+	//zap.S().Debugf("[HT][PV] 2")
+	{
+
+		var stmtCopy *sql.Stmt
+		stmtCopy, err = txn.Prepare(pq.CopyIn("tmp_processvaluetable64", "timestamp", "asset_id", "value", "valuename"))
+		if err != nil {
+			zap.S().Errorf("Error preparing copy statement: %s", err.Error())
+			return messages, err, true, "Error preparing copy statement"
+		}
+
+		//zap.S().Debugf("[HT][PV] 3 %d", len(messages))
+		// Copy into the temporary table
+		for _, message := range messages {
+			couldParse, parsedMessage := ParseMessage(message)
+			if !couldParse {
+				//zap.S().Debugf("[HT][PV] Could not parse message: %s", message.String())
+
+				continue
+			}
+
+			var sC processValue
+			err = jsoniter.Unmarshal(parsedMessage.Payload, &sC)
+			if err != nil {
+
+				//zap.S().Debugf("[HT][PV] Could not parse message: %s", err.Error())
+				continue
+			}
+			AssetTableID, success := GetAssetTableID(parsedMessage.CustomerId, parsedMessage.Location, parsedMessage.AssetId)
+			if !success {
+				zap.S().Errorf("Error getting asset table id: %s for %s %s %s", err.Error(), parsedMessage.CustomerId, parsedMessage.Location, parsedMessage.AssetId)
+				putBackMsg = append(putBackMsg, message)
+				continue
+			}
+			timestampString, timestampInParsedMessagePayload := sC["timestamp_ms"]
+			if !timestampInParsedMessagePayload {
+
+				//zap.S().Debugf("[HT][PV] Timestamp not in parsed message payload")
+				continue
+			}
+
+			tsF64 := timestampString.(float64)
+			timestampMs := uint64(tsF64)
+			////zap.S().Debugf("[HT][PV] Timestamp: %d", timestampMs)
+			for k, v := range sC {
+				switch k {
+				case "timestamp_ms":
+				// Copied these exceptions from mqtt-to-postgresql
+				case "measurement":
+				case "serial_number":
+					break
+				default:
+					////zap.S().Debugf("[HT][PV] Inserting %s: %v", k, v)
+					value, valueIsFloat64 := v.(float64)
```

Comment:
see type conversion comment further above

Reply:
This is safe, it will return a boolean if it can't be casted.
https://mussatto.github.io/golang/go/safe/cast/2016/10/22/golang-safe-cast.html